server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -66,6 +67,8 @@ private static FieldFetcher create(
         Map<String, FieldContext> fieldContexts = new LinkedHashMap<>();
         List<String> unmappedFetchPattern = new ArrayList<>();
 
+        Map<String, LookupFieldCollector> lookupFieldCollectors = new HashMap<>();
```

Comment:
Here's an idea for how to refactor this so we don't need this new `LookupFieldCollector` abstraction:
* Update `ValueFetcher` with a new method `fetchDocumentField` that returns `DocumentField` and can be overridden by `LookupRuntimeFieldType`. Use this new method in `FieldFetcher` instead of `fetchValues`. Include a default implementation that delegates to `fetchValues` so we don't need to update a bunch of classes that already implement `fetchValues`.
* Make `LookupRuntimeFieldType#valueFetcher` be responsible for fetching its own information about the input field values, instead of doing it here in `FieldFetcher`. I think that it could just create its own value fetcher and perform the lookup itself.

I didn't try this out in detail so let me know if I'm missing something 🤔

Reply:
I tried to avoid looking up input fields twice when they are requested in the fetch fields, but it seems we cache them. I integrated your suggestion in https://github.com/elastic/elasticsearch/pull/82385/commits/bb278b995812e71b55d8bd9abbe47b3f5d51bb14.